R/template.R

```diff
@@ -266,133 +265,161 @@ prob_estimate_convert <- function(estimate) {
-#' class that `truth` and `estimate` should be, respectively. If `truth` and
-#' `estimate` are of the same class, just supply a vector of length 1. If
-#' they are different, supply a vector of length 2. For matrices, it is best
-#' to supply `"numeric"` as the class to check here.
-#'
-#' @param estimator The type of averaging to use. By this point, the averaging
-#' type should be finalized, so this should be a character vector of length 1\.
-#' By default, this character value is required to be one of: `"binary"`,
-#' `"macro"`, `"micro"`, or `"macro_weighted"`. If your metric allows more
-#' or less averaging methods, override this with `averaging_override`.
-#'
-#' @param case_weights Optionally, the realized case weights, as a numeric
-#' vector. This must be the same length as `truth`, and will be considered in
-#' the `na_rm` checks. If supplied, this will be passed on to `metric_impl` as
-#' the named argument `case_weights`.
+#' @param case_weights the realized case weights, as a numeric vector. This must
+#'   be the same length as `truth`, and will be considered in the `na_rm`
+#'   checks. If supplied, this will be passed on to `metric_impl` as the named
+#'   argument `case_weights`.
 #'
 #' @param ... Extra arguments to your core metric function, `metric_impl`, can
-#' technically be passed here, but generally the extra args are added through
-#' R's scoping rules because the core metric function is created on the fly
-#' when the vector method is called.
+#'   technically be passed here, but generally the extra args are added through
+#'   R's scoping rules because the core metric function is created on the fly
+#'   when the vector method is called.
 #'
 #' @details
 #'
-#' `metric_vec_template()` is called from the vector implementation of your
-#' metric. Also defined inside your vector implementation is a separate
+#' `class_metric_vec_template()`, `numeric_metric_vec_template()`, and
+#' `prob_metric_vec_template()` are called from the vector implementation of
+#' your metric. Also defined inside your vector implementation is a separate
 #' function performing the core implementation of the metric function. This
-#' core function is passed along to `metric_vec_template()` as `metric_impl`.
+#' core function is passed along as `metric_impl`.
 #'
-#' @seealso [metric_summarizer()] [finalize_estimator()] [dots_to_estimate()]
+#' @seealso [metric-summarizers()] [finalize_estimator()] [dots_to_estimate()]
 #'
+#' @name metric-vec_template
+NULL
+
+#' @rdname metric-vec_template
 #' @export
-metric_vec_template <- function(metric_impl,
-                                truth,
-                                estimate,
-                                na_rm = TRUE,
-                                cls = "numeric",
-                                estimator = NULL,
-                                case_weights = NULL,
-                                ...) {
+numeric_metric_vec_template <- function(metric_impl,
```

Comment:
It is interesting to me that the vec template functions now have 4 purposes:

- Validate case weights
- Convert from class_pred->factor as needed
- Handle missing values. Importantly, return `NA_real_` as needed rather than calling the `metric_impl`.
- Call the `metric_impl()`

The first 2 could easily be wrapped up in helpers that we'd expect the metric functions to call themselves (you already have `validate_case_weights()` and the other could be `truth <- convert_class_pred(truth)`)

The 3rd could be put into its own helper too, basically `handle_na_rm()` that I already mentioned, but this is a little trickier because the return value is somewhat complicated and the metric functions would have to correctly return `NA_real_` themselves if `handle_na_rm()$return_missing` was true

---

There are a few really big upsides of this though:
- You wouldn't need these template functions _at all_. You'd put slightly more burden on the metric authors to call the `validate_*()` and `handle_na_rm()` functions manually, but you'd remove a ton of pain related to having to call the `metric_impl()` function for them.
- Metric function authors wouldn't need to do the weird function-within-function thing. Like defining `j_index_impl()` inside `j_index_vec()`. If they are in charge of calling `j_index_impl()`, it can be put anywhere and could have `estimator, event_level` as arguments that could be manually passed through. That would be way simpler and easier to explain.
- It would also be way more flexible because metric authors wouldn't be boxed in by our validation checking code. If they don't want it or need something else, they can just roll their own

It might be worth trying the idea out on a few metrics to see if it would be worth it. I think the toughest part is just handling `handle_na_rm()`, but I think the pattern used in every metric would be

```r
res <- handle_na_rm(truth, estimate, case_weights, na_rm)
truth <- res$truth
estimate <- res$estimate
case_weights <- res$case_weights

if (res$return_missing) {
  return NA_real_
}

j_index_impl(truth, estimate, case_weights, estimator, event_level)
```

which isn't really that bad